backend/src/controllers/UserController.ts

```diff
@@ -155,4 +171,143 @@ async function getPermissions(user: Doc<UserModel>): Promise<Permission[]> {
+        });
+        return;
+    }
+
+    // Get the user
+    const user = await User.findById(req.params.userId);
+    if (!user) {
+        res.status(400).json({
+            status: 'error',
+            message: 'User not found',
+        });
+        return;
+    }
+
+    const addedRoles: RoleModel[] = [];
+
+    for (const roleId of req.body.roles) {
+        // Check if role exists
+        const role = await Role.findById(roleId);
+        if (!role) {
+            res.status(400).json({
+                status: 'error',
+                message: 'Role not found',
+            });
+            return;
+        }
+
+        // Check if the user already has the role
+        if (user.roles.includes(role.id)) {
+            res.status(400).json({
+                status: 'error',
+                message: 'User already has that role',
+            });
+            return;
+        }
+
+        // Check if trying to add the same role twice
+        if (addedRoles.includes(role.id)) {
+            res.status(400).json({
+                status: 'error',
+                message: 'You cannot add the same role twice',
+            });
+            return;
+        }
+
+        addedRoles.push(role);
+    }
+
+    // Give the roles to the user
+    for (const role of addedRoles) {
+        user.roles.push(role);
+    }
+
+    await user.save();
+
+    res.status(200).json({
+        status: 'success',
+        rolesAdded: addedRoles.length,
+        data: {
+            user,
+            addedRoles,
+        },
+    });
+});
+
+/**
+ * @desc 	Removes roles specified in request body from a user
+ * @route 	DELETE /api/users/:userId/roles/
+ */
+
+const removeRoles = asyncHandler(async (req: TypedRequest<{ roles: string[] }, UserIdParam>, res: Response) => {
+    // Check atleast one role specified
+    if (req.body.roles.length == 0) {
+        res.status(400).json({
+            status: 'error',
+            message: 'You must specify atleast one role',
+        });
+        return;
+    }
+
+    // Get the user
+    const user = await User.findById(req.params.userId).populate('roles');
+    if (!user) {
+        res.status(400).json({
+            status: 'error',
+            message: 'User not found',
+        });
+        return;
+    }
+
+    const removedRoles: RoleModel[] = [];
+
+    for (const roleId of req.body.roles) {
+        // Check if role exists
+        const role = await Role.findById(roleId);
+        if (!role) {
+            res.status(400).json({
+                status: 'error',
+                message: 'Role not found',
+            });
+            return;
+        }
+
+        removedRoles.push(role);
+    }
+
+    // Remove the roles from the user or the removed list if the user doesn't have the role
+    for (const role of removedRoles) {
+        const index = user.roles.findIndex((r) => r.name == role.name);
+        if (index >= 0) {
+            user.roles.splice(index, 1);
```

Comment:
I think you should `findIndex` based on the `_id` rather than the name

Reply:
`const index = user.roles.findIndex((r) => r._id== role._id);` doesn't work... any idea's what I'm missing?

Reply:
@lpar105 It turns out when comparing `ObjectId`s you're meant to use the `.equals` method, so it should be:
```ts
r._id.equals(role._id)
```